documents/entityrelationship/service.go

```diff
@@ -190,33 +190,93 @@ func (s service) DeriveEntityRelationshipResponse(model documents.Model) (*entit
 
 }
 
-// DeriveEntityData returns create response from entity model
-func (s service) DeriveEntityRelationshipData(doc documents.Model) (*entitypb.RelationshipData, error) {
-	er, ok := doc.(*EntityRelationship)
+// DeriveEntityRelationshipData returns create response from entity model
+func (s service) DeriveEntityRelationshipData(model documents.Model) (*entitypb.RelationshipData, error) {
+	er, ok := model.(*EntityRelationship)
 	if !ok {
 		return nil, documents.ErrDocumentInvalidType
 	}
 
-	return er.getClientData(), nil
+	return er.getRelationshipData(), nil
 }
 
-// DeriveFromUpdatePayload returns a new version of the old entity identified by identifier in payload
+// DeriveFromUpdatePayload returns a new version of the indicated Entity Relationship with a deleted access token
 func (s service) DeriveFromUpdatePayload(ctx context.Context, payload *entitypb.RelationshipPayload) (documents.Model, error) {
 	if payload == nil {
-		return nil, documents.ErrDocumentNil
+		return nil, documents.ErrPayloadNil
+	}
+
+	eID, err := hexutil.Decode(payload.Identifier)
+	if err != nil {
+		return nil, err
+	}
+
+	did, err := identity.StringsToDIDs(payload.TargetIdentity)
+	if err != nil {
+		return nil, err
+	}
+
+	selfDID, err := contextutil.AccountDID(ctx)
+	if err != nil {
+		return nil, errors.New("failed to get self ID")
+	}
+
+	id, err := s.repo.FindEntityRelationshipIdentifier(eID, selfDID, *did[0])
+	if err != nil {
+		return nil, err
+	}
+
+	r, err := s.GetCurrentVersion(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+
+	model, err := r.(*EntityRelationship).DeleteAccessToken(ctx, hexutil.Encode(did[0][:]))
+	if err != nil {
+		return nil, err
 	}
-	return nil, nil
-	// get latest old version of the document
-}
 
-// Get returns the entity relationship requested
-// TODO
-func (s service) Get(ctx context.Context, payload *entitypb.RelationshipData) (documents.Model, error) {
-	return nil, nil
+	r.(*EntityRelationship).Document = model.Document
+	return r, nil
 }
 
-// GetEntityRelation returns a entity relation based on an entity id
-func (s service) GetEntityRelation(entityIdentifier, version []byte) (*EntityRelationship, error) {
-	// todo not implemented yet
-	return nil, nil
+// Get returns the requested version of the entity relationships that involve the entityID passed in
+// If version is nil, returns the latest version of the relationship
+func (s service) GetEntityRelationships(ctx context.Context, entityID, version []byte) ([]EntityRelationship, error) {
+	var relationships []EntityRelationship
+	if entityID == nil {
+		return nil, documents.ErrPayloadNil
+	}
+
+	selfDID, err := contextutil.AccountDID(ctx)
+	if err != nil {
+		return nil, errors.New("failed to get self ID")
+	}
+
+	relevant, err := s.repo.ListAllRelationships(entityID, selfDID)
+	if err != nil {
+		return nil, err
+	}
+
+	if version != nil {
+		for _, v := range relevant {
+			r, err := s.GetVersion(ctx, v, version)
```

Comment:
Let say I have 3 ER.
Every ER would have a different `current_version` identifier. GetVersion would return the document where `current_version == version`.  It would only work for one ER but would fail for the two others.

We currently don't have a method where you could say return version number 1 or 2.
Do we need the version part at all in the method?

Reply:
The db method ListAllEntityRelationships lists unique relationships, so this shouldn't happen.

Reply:
let's discuss this